File: rust/src/vnd.rs
// SPDX-FileCopyrightText: 2026 Arnaldo Mendes Pires Junior
// SPDX-License-Identifier: MIT

use crate::cache::EvaluationCache;
use crate::grasp::evaluate_with_cache;
use crate::helpers::{clamp, expired, normalize_integer_tail};
use rand::Rng;
use rand_chacha::ChaCha8Rng;
use std::time::Instant;

/// Try integer moves for a single variable: base-1, base, base+1.
fn try_integer_moves<F>(
    idx: usize,
    solution: &mut Vec<f64>,
    best_cost: f64,
    func: &F,
    lower: &[f64],
    upper: &[f64],
    cache: &mut Option<EvaluationCache>,
    half: usize,
) -> (f64, bool)
where
    F: Fn(&[f64]) -> f64,
{
    let base = solution[idx].round();
    let mut best = best_cost;
    let mut improved = false;

    for delta in [-1.0, 0.0, 1.0] {
        let val = clamp(base + delta, lower[idx], upper[idx]).round();
        if (val - solution[idx]).abs() < 1e-12 {
            continue;
        }
        let old = solution[idx];
        solution[idx] = val;
        let cost = evaluate_with_cache(solution, func, cache, half);
        if cost < best {
            best = cost;
            improved = true;
        } else {
            solution[idx] = old;
        }
    }
    (best, improved)
}

/// Try a continuous move: ±5% of span.
fn try_continuous_move<F>(
    idx: usize,
    solution: &mut Vec<f64>,
    best_cost: f64,
    func: &F,
    rng: &mut ChaCha8Rng,
    lower: &[f64],
    upper: &[f64],
    cache: &mut Option<EvaluationCache>,
    half: usize,
) -> (f64, bool)
where
    F: Fn(&[f64]) -> f64,
{
    let span = upper[idx] - lower[idx];
    let delta = rng.random_range(-0.05..=0.05) * span;
    let new_val = clamp(solution[idx] + delta, lower[idx], upper[idx]);
    let old = solution[idx];
    solution[idx] = new_val;
    let cost = evaluate_with_cache(solution, func, cache, half);
    if cost < best_cost {
        (cost, true)
    } else {
        solution[idx] = old;
        (best_cost, false)
    }
}

/// Perturb a single index (for swap/multiflip neighborhoods).
fn perturb_index(
    solution: &mut [f64],
    idx: usize,
    strength: usize,
    rng: &mut ChaCha8Rng,
    lower: &[f64],
    upper: &[f64],
    half: usize,
) {
    if idx >= half {
        let step = (strength as f64 / 2.0).max(1.0);
        let delta = rng.random_range(-step..=step);
        solution[idx] = clamp((solution[idx] + delta).round(), lower[idx], upper[idx]);
    } else {
        let span = upper[idx] - lower[idx];
        let delta = rng.random_range(-0.15..=0.15) * span;
        solution[idx] = clamp(solution[idx] + delta, lower[idx], upper[idx]);
    }
}

/// Neighborhood 1: single-variable flip (prioritized by sensitivity).
fn neighborhood_flip<F>(
    solution: &mut Vec<f64>,
    best_cost: f64,
    func: &F,
    sensitivity: &mut [f64],
    rng: &mut ChaCha8Rng,
    lower: &[f64],
    upper: &[f64],
    cache: &mut Option<EvaluationCache>,
    half: usize,
    deadline: Option<Instant>,
) -> (f64, bool)
where
    F: Fn(&[f64]) -> f64,
{
    let n = solution.len();
    let mut indices: Vec<usize> = (0..n).collect();
    indices.sort_by(|&a, &b| sensitivity[b].partial_cmp(&sensitivity[a]).unwrap());

    let mut current_best = best_cost;
    let mut any_improved = false;

    for &idx in &indices {
        if expired(deadline) {
            break;
        }
        let (new_cost, improved) = if idx >= half {
            try_integer_moves(idx, solution, current_best, func, lower, upper, cache, half)
        } else {
            try_continuous_move(
                idx,
                solution,
                current_best,
                func,
                rng,
                lower,
                upper,
                cache,
                half,
            )
        };
        if improved {
            let delta = current_best - new_cost;
            sensitivity[idx] = 0.9 * sensitivity[idx] + delta;
            current_best = new_cost;
            any_improved = true;
        }
    }
    (current_best, any_improved)
}

/// Neighborhood 2: paired variable swap.
fn neighborhood_swap<F>(
    solution: &mut Vec<f64>,
    best_cost: f64,
    func: &F,
    rng: &mut ChaCha8Rng,
    lower: &[f64],
    upper: &[f64],
    cache: &mut Option<EvaluationCache>,
    half: usize,
    deadline: Option<Instant>,
) -> (f64, bool)
where
    F: Fn(&[f64]) -> f64,
{
    let n = solution.len();
    let mut current_best = best_cost;
    let mut any_improved = false;
    let max_attempts = 50.min(n * (n - 1) / 2);

    for _ in 0..max_attempts {
        if expired(deadline) {
            break;
        }
        let i = rng.random_range(0..n);
        let j = rng.random_range(0..n);
        if i == j {
            continue;
        }
        let old_i = solution[i];
        let old_j = solution[j];
        perturb_index(solution, i, 4, rng, lower, upper, half);
        perturb_index(solution, j, 4, rng, lower, upper, half);
        normalize_integer_tail(solution, half);
        let cost = evaluate_with_cache(solution, func, cache, half);
        if cost < current_best {
            current_best = cost;
            any_improved = true;
        } else {
            solution[i] = old_i;
            solution[j] = old_j;
        }
    }
    (current_best, any_improved)
}

/// Neighborhood 3: simultaneous k-variable changes.
fn neighborhood_multiflip<F>(
    solution: &mut Vec<f64>,
    best_cost: f64,
    func: &F,
    rng: &mut ChaCha8Rng,
    lower: &[f64],
    upper: &[f64],
    cache: &mut Option<EvaluationCache>,
    half: usize,
    deadline: Option<Instant>,
) -> (f64, bool)
where
    F: Fn(&[f64]) -> f64,
{
    let n = solution.len();
    let k = 3.min(n);
    let mut current_best = best_cost;
    let mut any_improved = false;

    for _ in 0..50 {
        if expired(deadline) {
            break;
        }
        let backup = solution.to_vec();
        let mut indices: Vec<usize> = (0..n).collect();
        // Fisher-Yates partial shuffle for k elements
        for i in 0..k {
            let j = rng.random_range(i..n);
            indices.swap(i, j);
        }
        for &idx in &indices[..k] {
            perturb_index(solution, idx, 4, rng, lower, upper, half);
        }
        normalize_integer_tail(solution, half);
        let cost = evaluate_with_cache(solution, func, cache, half);
        if cost < current_best {
            current_best = cost;
            any_improved = true;
        } else {
            solution.copy_from_slice(&backup);
        }
    }
    (current_best, any_improved)
}

/// VND: Variable Neighborhood Descent.
pub(crate) fn local_search_vnd<F>(
    func: &F,
    solution: &mut Vec<f64>,
    current_cost: f64,
    half: usize,
    lower: &[f64],
    upper: &[f64],
    max_iter: usize,
    cache: &mut Option<EvaluationCache>,
    rng: &mut ChaCha8Rng,
    deadline: Option<Instant>,
) -> f64
where
    F: Fn(&[f64]) -> f64,
{
    let n = solution.len();
    let mut sensitivity = vec![0.0_f64; n];
    let mut best_cost = current_cost;
    let no_improve_limit = 5;
    let mut no_improve_flip_count = 0;

    for _ in 0..max_iter {
        if expired(deadline) {
            break;
        }

        // Decay sensitivity
        for s in sensitivity.iter_mut() {
            *s *= 0.9;
        }

        // Try flip neighborhood
        let (cost, improved) = neighborhood_flip(
            solution,
            best_cost,
            func,
            &mut sensitivity,
            rng,
            lower,
            upper,
            cache,
            half,
            deadline,
        );
        best_cost = cost;

        if !improved {
            no_improve_flip_count += 1;

            // Try swap neighborhood
            let (cost2, improved2) = neighborhood_swap(
                solution, best_cost, func, rng, lower, upper, cache, half, deadline,
            );
            best_cost = cost2;

            if !improved2 && no_improve_flip_count >= no_improve_limit {
                // Try multiflip neighborhood
                let (cost3, _) = neighborhood_multiflip(
                    solution, best_cost, func, rng, lower, upper, cache, half, deadline,
                );
                best_cost = cost3;
            }

            if no_improve_flip_count >= no_improve_limit {
                break;
            }
        } else {
            no_improve_flip_count = 0;
        }
    }
    best_cost
}

File: julia/src/vnd.jl
# SPDX-FileCopyrightText: 2026 Arnaldo Mendes Pires Junior
# SPDX-License-Identifier: MIT

"""VND — Variable Neighborhood Descent."""

function create_cached_cost_fn(
    cost_fn::Function,
    cache::Union{EvaluationCache, Nothing},
)::Function
    function cached_cost_fn(sol::Vector{Float64})::Float64
        if cache !== nothing
            cached = cache_get(cache, sol)
            cached !== nothing && return cached
        end
        cost = cost_fn(sol)
        if cache !== nothing
            cache_put!(cache, sol, cost)
        end
        return cost
    end
    return cached_cost_fn
end

# --- Atomic move helpers ---

function try_integer_moves(
    idx::Int,
    sol::Vector{Float64},
    best_benefit::Float64,
    cost_fn::Function,
    lower::Union{Vector{Float64}, Nothing},
    upper::Union{Vector{Float64}, Nothing},
)
    old = sol[idx]
    base = round(Int, old)
    cand_vals = (base - 1, base, base + 1)
    lo = lower !== nothing ? ceil(Int, lower[idx]) : typemin(Int)
    hi = upper !== nothing ? floor(Int, upper[idx]) : typemax(Int)
    for v in cand_vals
        (v < lo || v > hi) && continue
        sol[idx] = Float64(v)
        c = cost_fn(sol)
        if c < best_benefit
            result = copy(sol)
            return result, c, true
        end
    end
    sol[idx] = old
    return sol, best_benefit, false
end

function try_continuous_move(
    idx::Int,
    sol::Vector{Float64},
    best_benefit::Float64,
    cost_fn::Function,
    rng::AbstractRNG,
    lower::Union{Vector{Float64}, Nothing},
    upper::Union{Vector{Float64}, Nothing},
)
    old = sol[idx]
    if lower !== nothing && upper !== nothing
        span = upper[idx] - lower[idx]
        perturb = (rand(rng) - 0.5) * 0.1 * span  # ±5%
    else
        perturb = (rand(rng) - 0.5) * 0.2
    end
    new_val = old + perturb
    if lower !== nothing && upper !== nothing
        new_val = clamp(new_val, lower[idx], upper[idx])
    end
    sol[idx] = new_val
    cost = cost_fn(sol)
    if cost < best_benefit
        return true, cost
    end
    sol[idx] = old
    return false, best_benefit
end

function perturb_index!(
    perturbed::Vector{Float64},
    idx::Int,
    strength::Int,
    rng::AbstractRNG,
    lower::Union{Vector{Float64}, Nothing},
    upper::Union{Vector{Float64}, Nothing},
)
    half = get_half(length(perturbed))
    old = perturbed[idx]
    if idx > half  # integer variable (1-indexed: > half means integer part)
        lo = lower !== nothing ? ceil(Int, lower[idx]) : nothing
        hi = upper !== nothing ? floor(Int, upper[idx]) : nothing
        step = max(1, round(Int, strength / 2))
        delta = rand(rng, (-step):step)
        new_val = round(Int, old) + delta
        if lo !== nothing && hi !== nothing
            new_val = clamp(new_val, lo, hi)
        end
        perturbed[idx] = Float64(new_val)
    else  # continuous variable
        if lower !== nothing && upper !== nothing
            span = upper[idx] - lower[idx]
            delta = (rand(rng) - 0.5) * 0.3 * span  # ±15%
            perturbed[idx] = clamp(old + delta, lower[idx], upper[idx])
        else
            delta = randn(rng) * 0.12 * (abs(old) + 1e-6)
            perturbed[idx] = old + delta
        end
    end
end

# --- Neighbourhood functions ---

function neighborhood_flip(
    cost_fn::Function,
    solution::Vector{Float64},
    current_benefit::Float64,
    num_vars::Int;
    first_improvement::Bool = true,
    seed::Union{Int, Nothing} = nothing,
    lower::Union{Vector{Float64}, Nothing} = nothing,
    upper::Union{Vector{Float64}, Nothing} = nothing,
    sensitivity::Union{Vector{Float64}, Nothing} = nothing,
    deadline::Float64 = 0.0,
)
    rng = new_rng(seed)
    half = get_half(num_vars)

    if sensitivity !== nothing && any(sensitivity .> 0)
        noise = rand(rng, num_vars) .* 0.1 .* maximum(sensitivity)
        priority = sensitivity .+ noise
        indices = sortperm(priority; rev = true)
    else
        indices = randperm(rng, num_vars)
    end

    int_indices = filter(i -> i > half, indices)
    cont_indices = filter(i -> i <= half, indices)

    best_solution = copy(solution)
    best_benefit = current_benefit

    # Integer flip
    for (count, i) in enumerate(int_indices)
        count % 8 == 0 && expired(deadline) && break
        new_sol, new_ben, improved =
            try_integer_moves(i, solution, best_benefit, cost_fn, lower, upper)
        if improved
            best_benefit = new_ben
            best_solution = copy(new_sol)
            first_improvement && return best_solution, best_benefit
        end
    end

    # Continuous flip
    for (count, i) in enumerate(cont_indices)
        count % 8 == 0 && expired(deadline) && break
        old_val = solution[i]
        changed, new_ben =
            try_continuous_move(i, solution, best_benefit, cost_fn, rng, lower, upper)
        if changed
            best_benefit = new_ben
            best_solution = copy(solution)
            first_improvement && return best_solution, best_benefit
        end
        solution[i] = old_val
    end

    return best_solution, best_benefit
end

function neighborhood_swap(
    cost_fn::Function,
    solution::Vector{Float64},
    current_benefit::Float64,
    num_vars::Int;
    first_improvement::Bool = true,
    max_attempts::Int = 50,
    lower::Union{Vector{Float64}, Nothing} = nothing,
    upper::Union{Vector{Float64}, Nothing} = nothing,
    deadline::Float64 = 0.0,
)
    best_solution = copy(solution)
    best_benefit = current_benefit
    rng = new_rng()
    half = get_half(num_vars)
    (half <= 0 || half >= num_vars) && return best_solution, best_benefit
    n_int = num_vars - half

    for _ in 1:max_attempts
        expired(deadline) && break
        cont_idx = rand(rng, 1:half)
        int_idx = half + rand(rng, 1:n_int)

        old_cont = solution[cont_idx]
        old_int = solution[int_idx]

        if lower !== nothing && upper !== nothing
            span_cont = upper[cont_idx] - lower[cont_idx]
            solution[cont_idx] = clamp(
                old_cont + (rand(rng) - 0.5) * 0.16 * span_cont,
                lower[cont_idx],
                upper[cont_idx],
            )
            lo_int = ceil(Int, lower[int_idx])
            hi_int = floor(Int, upper[int_idx])
            new_int = round(Int, old_int) + rand(rng, -1:1)
            solution[int_idx] = Float64(clamp(new_int, lo_int, hi_int))
        else
            solution[cont_idx] = old_cont + (rand(rng) - 0.5) * 0.2
            solution[int_idx] = Float64(round(Int, old_int) + rand(rng, -1:1))
        end

        cost = cost_fn(solution)
        if cost < best_benefit
            best_benefit = cost
            best_solution = copy(solution)
            first_improvement && return best_solution, best_benefit
        end

        solution[cont_idx] = old_cont
        solution[int_idx] = old_int
    end

    return best_solution, best_benefit
end

function neighborhood_multiflip(
    cost_fn::Function,
    solution::Vector{Float64},
    current_benefit::Float64,
    num_vars::Int;
    k::Int = 3,
    max_attempts::Int = 50,
    seed::Union{Int, Nothing} = nothing,
    lower::Union{Vector{Float64}, Nothing} = nothing,
    upper::Union{Vector{Float64}, Nothing} = nothing,
    deadline::Float64 = 0.0,
)
    best_solution = copy(solution)
    best_benefit = current_benefit
    rng = new_rng(seed)
    half = get_half(num_vars)

    for _ in 1:max_attempts
        expired(deadline) && break
        indices = sort(randperm(rng, num_vars)[1:min(k, num_vars)])
        old_vals = solution[indices]

        # Modify integer indices
        for idx in indices
            if idx > half
                base = round(Int, solution[idx])
                delta = rand(rng, -1:1)
                new_val = base + delta
                if lower !== nothing && upper !== nothing
                    lo = ceil(Int, lower[idx])
                    hi = floor(Int, upper[idx])
                    new_val = clamp(new_val, lo, hi)
                end
                solution[idx] = Float64(new_val)
            else
                if lower !== nothing && upper !== nothing
                    span = upper[idx] - lower[idx]
                    perturb = (rand(rng) - 0.5) * 0.1 * span
                else
                    perturb = (rand(rng) - 0.5) * 0.2
                end
                new_val = solution[idx] + perturb
                if lower !== nothing && upper !== nothing
                    new_val = clamp(new_val, lower[idx], upper[idx])
                end
                solution[idx] = new_val
            end
        end

        cost = cost_fn(solution)
        if cost < best_benefit
            best_benefit = cost
            best_solution = copy(solution)
        end
        solution[indices] = old_vals
    end

    return best_solution, best_benefit
end

function try_neighborhoods(
    cached_cost_fn::Function,
    solution::Vector{Float64},
    current_benefit::Float64,
    num_vars::Int;
    first_improvement::Bool = true,
    iteration::Int = 1,
    no_improve_flip_limit::Int = 3,
    lower::Union{Vector{Float64}, Nothing} = nothing,
    upper::Union{Vector{Float64}, Nothing} = nothing,
    sensitivity::Union{Vector{Float64}, Nothing} = nothing,
    deadline::Float64 = 0.0,
)
    expired(deadline) && return solution, current_benefit, false

    new_sol, new_ben = neighborhood_flip(
        cached_cost_fn,
        solution,
        current_benefit,
        num_vars;
        first_improvement,
        lower,
        upper,
        sensitivity,
        deadline,
    )
    new_ben < current_benefit && return new_sol, new_ben, true

    expired(deadline) && return solution, current_benefit, false

    new_sol, new_ben = neighborhood_swap(
        cached_cost_fn,
        solution,
        current_benefit,
        num_vars;
        first_improvement,
        lower,
        upper,
        deadline,
    )
    new_ben < current_benefit && return new_sol, new_ben, true

    if iteration % no_improve_flip_limit == 0
        expired(deadline) && return solution, current_benefit, false
        new_sol, new_ben = neighborhood_multiflip(
            cached_cost_fn,
            solution,
            current_benefit,
            num_vars;
            k = no_improve_flip_limit,
            lower,
            upper,
            deadline,
        )
        new_ben < current_benefit && return new_sol, new_ben, true
    end

    return solution, current_benefit, false
end

function local_search_vnd(
    cost_fn::Function,
    solution::Vector{Float64},
    num_vars::Int;
    max_iter::Int = 300,
    first_improvement::Bool = true,
    no_improve_limit::Int = 5,
    no_improve_flip_limit::Int = 3,
    lower::Union{Vector{Float64}, Nothing} = nothing,
    upper::Union{Vector{Float64}, Nothing} = nothing,
    cache::Union{EvaluationCache, Nothing} = nothing,
    deadline::Float64 = 0.0,
)::Vector{Float64}
    solution = copy(solution)
    cached_cost_fn = create_cached_cost_fn(cost_fn, cache)
    current_benefit = cached_cost_fn(solution)

    sensitivity = zeros(Float64, num_vars)
    iteration = 0
    no_improve_count = 0

    while iteration < max_iter && no_improve_count < no_improve_limit
        expired(deadline) && break
        iteration += 1
        old_benefit = current_benefit
        old_solution = copy(solution)

        solution, current_benefit, improved = try_neighborhoods(
            cached_cost_fn,
            solution,
            current_benefit,
            num_vars;
            first_improvement,
            iteration,
            no_improve_flip_limit,
            lower,
            upper,
            sensitivity,
            deadline,
        )

        if improved
            no_improve_count = 0
            # Update sensitivity (P9)
            delta = abs(current_benefit - old_benefit)
            for i in 1:num_vars
                if solution[i] != old_solution[i]
                    sensitivity[i] = sensitivity[i] * 0.9 + delta
                else
                    sensitivity[i] *= 0.9
                end
            end
        else
            no_improve_count += 1
            sensitivity .*= 0.9
        end
    end

    return solution
end
